(superseded) llama : add llama_batch_ext

ngxson opened the pull request:
Ref comment: https://github.com/ggerganov/llama.cpp/pull/11292#discussion_r1946287043

Closes https://github.com/ggerganov/llama.cpp/issues/10381

Migration patterns:

```cpp
llama_batch batch = llama_batch_init(n_kv_max, 0, 1);
// becomes:
llama_batch_ext_ptr batch = llama_batch_ext_ptr(llama_batch_ext_init(n_kv_max, 1));


common_batch_add(batch, tokens[i], pos, { 0 }, false);
// becomes:
const llama_seq_id seq_id = 0;
llama_batch_ext_add_text(batch.get(), tokens[i], pos, &seq_id, 1, false);


llama_decode(lctx, llama_batch_get_one(tokens.data(), std::min(tokens.size(), (size_t) params.n_batch)));
// becomes:
llama_batch_ext_ptr batch(llama_batch_ext_init_from_text(tokens.data(), tokens.size(), 0, 0));
llama_decode_ext(lctx, batch.get());


llama_decode(ctx, batch);
// becomes:
llama_decode_ext(ctx, batch.get());
```

Current status:
- This PR currently contains the first proposal of public API that allows hiding `llama_batch` from public API --> To be discussed
- Only `llama-server` works for now
- TODO: the members of `llama_batch` can be migrated to cpp types

ngxson committed:
first proposal for private llama_batch

ngxson committed:
rework, targeting llama-server

ngxson commented:
@ggerganov Would you mind having a look on this initial proposal? Thank you!

ggerganov commented on src/llama-batch.h:
```diff
@@ -5,6 +5,32 @@
+// - logits : if zero, the logits (and/or the embeddings) for the respective token will not be output
+//            (if set to NULL, only the logits for last token will be returned)
+//
+struct llama_batch {
+    int32_t n_tokens;
+    int32_t max_tokens;
+    bool is_view;
+
+    llama_token  *  token;
+    float        *  embd;
+    llama_pos    *  pos;
+    int32_t      *  n_seq_id;
+    llama_seq_id ** seq_id;
+    int8_t       *  logits; // TODO: rename this to "output"
+};
+
```
This is overall looking good. Two things that come to mind:

- Should we try to make the change backwards compatible by adding `llama_batch_ext` and keeping `llama_batch` + deprecation notices? Later on, we remove `llama_batch`, but we have given developers time to adapt to `llama_batch_ext`. If we decide to do that, we would need to implement a small shim for `llama_batch_ext <-> llama_batch`.

- The new `llama_batch`, now that it is in the C++ world, can become more general-purpose and utilize STL containers. Of course, we should keep this PR simple and do this later - the main goal for now is just to remove the definition from the `llama.h` header so that we can later refactor more easily.

ngxson replied:
Ok so I tried adding the private batch API as `llama_batch_ext` and keeping the old `llama_batch` as-is. It's still quite messy for now, not yet working, but just to show you how it looks like.

One small change that I'm not 100% sure, the `llama_encode` and `llama_decode` functions both use `llama_batch`, and when moving to `_ext`, I need to change the function name to prevent `error: conflicting types`. Here I decide to change it to `llama_text_(en|de)code` to make it more aligned with `llama_vision_encode`. WDYT?

ggerganov replied:
> One small change that I'm not 100% sure, the llama_encode and llama_decode functions both use llama_batch, and when moving to _ext, I need to change the function name to prevent error: conflicting types. Here I decide to change it to llama_text_(en|de)code to make it more aligned with llama_vision_encode. WDYT?

Not convinced this is the best way. IMO the `llama_encode / llama_decode` should be oblivious to the contents of the batch. I am imagining in the future to add multi-modal `llama_context`s that would support simultaneously audio, vision, text, etc. So we should be able to add all of these types of inputs into a batch and call `llama_encode`. Internally the context will create the necessary sub-batches and process them with the respective compute graphs.

I think we can add `llama_encode_ext / llama_decode_ext` instead.

ngxson replied:
@ggerganov I've been able to adapt `common` to this new API. All `common_batch_*` will be soon removed.

For now, I established some migration patterns like:

```cpp
common_batch_add(batch, tokens[i], pos, { 0 }, false);
// becomes:
const llama_seq_id seq_id = 0;
llama_batch_ext_add_text_token(batch.get(), tokens[i], pos, &seq_id, 1, false);


llama_decode(lctx, llama_batch_get_one(tokens.data(), std::min(tokens.size(), (size_t) params.n_batch)));
// becomes:
llama_batch_ext_ptr batch(llama_batch_ext_init_from_text(tokens.data(), tokens.size(), 0, 0));
llama_encode_ext(lctx, batch.get());
```

The `llama_batch_ext_ptr` is a convenient cpp wrapper to free the memory automatically. However, I won't use this in certain examples including `simple` and `simple-chat`, to demo the usage with pure C API.

What do you think about this approach?

ngxson committed:
move to llama_batch_ext

ngxson committed:
server : use llama_batch_ext

ngxson committed:
fix server

ngxson committed:
llama_decode_ext

ngxson committed:
Merge branch 'master' into xsn/private_batch_api

ngxson committed:
adapt common

ngxson committed:
Merge branch 'master' into xsn/private_batch_api

ggerganov commented on include/llama.h:
```diff
@@ -256,6 +259,18 @@ extern "C" {
+    // It can contain text tokens and embeddings for one or many sequences
+    struct llama_batch_ext;
+
+    struct llama_batch_ext_token_info {
+        llama_token    token;
+        llama_pos      pos;
+        int32_t        n_seq_id;
+        llama_seq_id * seq_id;
+        int8_t         logits;
+    };
```
This might not be very future-proof. Mixed-modality batches would have tokens, embeddings and tensors mixed together in the same batch. So calling `llama_batch_ext_get_token_info(batch, i);` is not always well-defined because it might not be a token at position `i`.

Maybe we can postpone this "token_info" API. I think all usages in the examples that require to read back info from the batch can be implemented in the example code without relying on the API. This way we can focus only on implementing only the API for creating batches and adding data to them. Later on when we have a better idea of the implementation, we can add a helper API to get info back from the batches.

ngxson replied:
Yes I agree. Furthermore, this API requires doing a copy, so it won't be the best for performance. It's better to remove this API for now.

> I think all usages in the examples that require to read back info from the batch can be implemented in the example code without relying on the API.

This kind of logic is currently being used inside `llama-server`, not sure it appears on any other examples. I think I can make a thin wrapper for `llama_batch_ext` inside the example code. Feel free to tell me if you have a better idea.

ngxson replied:
This API is removed in https://github.com/ggml-org/llama.cpp/pull/11875/commits/1d6ba97789ed71a16b2e0b1d6f89dfda2499ac44 , a new `server_batch` wrapper is added to manage token logits placement in the batch

ggerganov commented on common/common.cpp:
```diff
@@ -1054,7 +1057,8 @@ struct common_init_result common_init_from_params(common_params & params) {
         if (llama_model_has_decoder(model)) {
-            llama_decode(lctx, llama_batch_get_one(tmp.data(), std::min(tmp.size(), (size_t) params.n_batch)));
+            llama_batch_ext_ptr batch(llama_batch_ext_init_from_text(tmp.data(), std::min(tmp.size(), (size_t) params.n_batch), 0, 0));
+            llama_encode_ext(lctx, batch.get());
```
This should be `llama_decode_ext`, correct?

ggerganov commented on include/llama.h:
```diff
@@ -842,29 +857,109 @@ extern "C" {
+    //  0 : success
+    // -1 : not enough space in the batch
+    // -2 : embd is already set, cannot add text tokens
+    LLAMA_API int32_t llama_batch_ext_add_text_token(
```
```suggestion
    LLAMA_API int32_t llama_batch_ext_add_text(
```

For consistency with `llama_batch_ext_init_from_text`. Othewise it would have to be `llama_batch_ext_init_from_text_tokens`.

ngxson committed:
correct llama_decode_ext

ngxson committed:
llama_batch_ext_add_text

ngxson committed:
remove token_info API

ngxson committed:
apply various in places

ngxson commented:
OK so I've been able to apply this to various example (not all of them). Would be nice if you can have a quick look @ggerganov before I migrate the rest.

One thing to note, the loop check over tokens in batch (discussed in https://github.com/ggml-org/llama.cpp/pull/11875#discussion_r1976395726) is used by both `server.cpp` and `embeddings.cpp`, so my solution was to create a thin wrapper called `common_batch`. Looks a bit messy for now, so I'm wondering if in the future we can have a `llama_get_embeddings_ext` or something that can make this easier.

ggerganov commented:
The `common_batch` is ok for now.

> Looks a bit messy for now, so I'm wondering if in the future we can have a llama_get_embeddings_ext or something that can make this easier.

It seems we rather need something to query the batch, no? How do you imagine `llama_get_embeddings_ext` to work?

I was thinking something like:

```c
struct llama_batch_ext_part;

llama_batch_ext_part * part = llama_batch_ext_get_part(batch, i);
if (llama_batch_ext_part_is_token(part)) {
    llama_token id = llama_batch_ext_part_get_id(part);
    ... get token id, sequence id, etc. ...
}
```

But since I'm not 100% about all the details yet related to multi-modal batches, I think it is better to postpone this API for later, and handle the batch information in the user code for now.

ngxson commented:
> How do you imagine `llama_get_embeddings_ext` to work?

I don't have a clear idea yet, but I'm thinking as a developer using `libllama` in the their program: Whenever I add a token to the batch, in case of text token I need to know:
- The token ID
- The pos
- The seq_id

So when I retrieve back the logits/embeddings, I would imagine that the `get_embeddings` function will have one of these 2 signatures:
- `get_embeddings(seq_id)` ==> we already had `llama_get_embeddings_seq`
- `get_embeddings(seq_id, pos)` ==> we currently need to read back the tokens from batch

--- 

> It seems we rather need something to query the batch, no?

Yes we can, and this will be quite similar to my point above. I'm thinking about these 2 options:
- Having something like `llama_batch_ext_query(seq_id, pos)` that returns the `output_id` of the token
- Or, explicitly has `llama_batch_ext_set_output(...)` that returns the `output_id`. That means the `logits` param will be removed from `llama_batch_ext_add_text`

ggerganov commented:
> Yes we can, and this will be quite similar to my point above. I'm thinking about these 2 options:

> Having something like llama_batch_ext_query(seq_id, pos) that returns the output_id of the token. This can then be used with llama_get_embeddings_ith(output_id)
Or, explicitly has llama_batch_ext_set_output(...) that returns the output_id. That means the logits param will be removed from llama_batch_ext_add_text
(Edit) Or, another option, llama_batch_ext_add_text can return the output_id if logits is set to true

Hm, yes. The `llama_batch_ext_set_output()` idea sounds good.

Btw, this makes me wonder if we should actually move the output buffers for logits and the embeddings to be owned by the `llama_batch_ext` (currently these buffers are owned by the `llama_context` and are shared by all batches)?

ggerganov commented:
Now that #12181 has been merged, it should be a good time to get this merged too.

ngxson commented:
Yes thanks for the heads up, I'll focus on finishing this today & tomorrow

ngxson committed:
Merge branch 'master' into xsn/private_batch_api

ngxson commented:
> Btw, this makes me wonder if we should actually move the output buffers for logits and the embeddings to be owned by the `llama_batch_ext` (currently these buffers are owned by the `llama_context` and are shared by all batches)?

If the output logits and embeddings are staying are `float *` or `std::vector<float>` then yes, I think it will be better to move them to `llama_batch_ext`

ngxson committed:
fix merge errors

ngxson committed:
return output ID from llama_batch_ext_add/set

ngxson commented:
Ok so what I've done is to allow "output ID" to be returned from `llama_batch_ext_add/set...`

```cpp
    // Add text tokens to the batch
    // Return values:
    // -1 : not enough space in the batch
    // -2 : embd is already set, cannot add text tokens
    // otherwise, returns the output ID
    LLAMA_API int32_t llama_batch_ext_add_text(
        struct llama_batch_ext * batch,
                   llama_token   token,
                     llama_pos   pos,
            const llama_seq_id * seq_ids,
                        size_t   n_seq_ids,
                          bool   output);

    // Set output (logits/embeddings) for the token in the ith sequence
    // If pos == -1, output will be set for the all tokens
    // Return values:
    // -1 : the token is not in the batch
    // otherwise, returns the output ID
    LLAMA_API int32_t llama_batch_ext_set_output(
        struct llama_batch_ext * batch,
                     llama_pos   pos,
                  llama_seq_id   seq_id);

    // Set output (logits/embeddings) for the last added token
    // Return values:
    // -1 : the batch is empty
    // otherwise, returns the output ID
    LLAMA_API int32_t llama_batch_ext_set_output_last(struct llama_batch_ext * batch);
```

After this is merged, we can start moving output to `llama_batch_ext` (we can finally call it "output" now instead of "logits"), by implementing these API:

```cpp
float * llama_batch_ext_get_logits(int32_t output_id);
float * llama_batch_ext_get_embeddings(int32_t output_id);
```

ggerganov commented:
> If the output logits and embeddings are staying are float * or std::vector<float> then yes, I think it will be better to move them to llama_batch_ext (and can be done in a follow-up PR)

Yes, definitely in a separate PR.

> Ok so what I've done is to allow "output ID" to be returned from llama_batch_ext_add/set...

Ok, just to make sure we are on the same page: currently the "output ID" is the same as the position of the token in the batch, which works. But the output id can actually be more general than a position - it is a unique identifier associated with a requested output. And internally, the batch would keep a map about what output information should be associated with each output id. Does that make sense?

ngxson commented:
> But the output id can actually be more general than a position - it is a unique identifier associated with a requested output. And internally, the batch would keep a map about what output information should be associated with each output id. Does that make sense?

Yes that is exactly what I have in mind. For now, the output_id increase by one for each token even if the added token does not have output.

But in the future it may make more sense to only increase output_id based on an internal mapping as you said. Ofc that will depend on the implementation, but my point is that output_id will no longer necessarily be the token position in batch

ngxson commented:
Ok I have been able to apply this to the rest of example, with some notes:
- [ ] `qwen2vl-cli` set custom position for each token, so I will need to add support for this case in `llama_batch_ext_init_from_embd` (for now, I remove the deprecation `llama_decode` just to make it work)
- [ ] `llama-android` is still not yet migrated
- [ ] `speculative.cpp` in the examples is broken for now, but I have too little knowledge to fix it. Could you please help? @ggerganov (I temporary add `#ifdef 0` to make it compile)
- [ ] I'm not sure if `swiftui` also need to be updated
- [ ] I should also find an AI "reviewer" that works with cpp, I feel like this kind of task (repetitive) is perfectly suitable for AI ; unfortunately @copilot does not work as we already known

ngxson committed:
apply to the rest

ngxson committed:
fix common_batch missing seq_id

ngxson committed:
compile ok

ngxson committed:
fix llama_batch_ext_init_from_text

ngxson commented:
Small note: to make the usage of `llama_batch_ext_init_from_text` easier, I added an argument `bool output_last` that internally enable output for last token in the input.

For the future `llama_batch_ext_get_logits` API, we may need to allow `output_id == -1` to get logits of last input (used in conjunction with `bool output_last` above). This can make the usage a bit more simple for basic text generation use case:

```cpp
llama_batch_ext_ptr batch(llama_batch_ext_init_from_text(inp.data(), inp.size() - 1, 0, 0, true));
llama_decode_ext(batch);
float * logits = llama_batch_ext_get_logits(-1);
```

ngxson committed:
rm redundant llama_batch_ext_set_output_last

ngxson commented on examples/run/run.cpp:
```diff
@@ -969,7 +969,7 @@ static int generate(LlamaData & llama_data, const std::string & prompt, std::str
 
         // prepare the next batch with the sampled token
-        batch = llama_batch_get_one(&new_token_id, 1);
+        batch.reset(llama_batch_ext_init_from_text(&new_token_id, 1, 0, 0, true));
```
Ok I forgot to keep track of n_past here

ngxson commented on examples/llama-bench/llama-bench.cpp:
```diff
@@ -1461,7 +1462,8 @@ static void test_gen(llama_context * ctx, int n_gen, int n_threads) {
 
     for (int i = 0; i < n_gen; i++) {
-        llama_decode(ctx, llama_batch_get_one(&token, 1));
+        llama_batch_ext_ptr batch(llama_batch_ext_init_from_text(&token, 1, 0, 0, true));
```
Maybe we can have a small wrapper to make this shorter:

```
llama_batch_ext_ptr::from_text(&token, 1, 0, 0, true);
```

ngxson commented on examples/llava/gemma3-cli.cpp:
```diff
@@ -87,50 +88,18 @@ struct gemma3_context {
     for (llama_token & t : tokens) {
-        common_batch_add(ctx.batch, t, ctx.n_past++, {0}, false);
+        llama_seq_id seq_id = 0;
+        llama_batch_ext_add_text(ctx.batch.get(), t, 0, &seq_id, 1, false);
```
```suggestion
        llama_batch_ext_add_text(ctx.batch.get(), t, ctx.n_past++, &seq_id, 1, false);
```

ngxson commented:
@coderabbitai review

ngxson commented:
@coderabbitai review

ngxson committed:
correct comment

ngxson committed:
bring back mistakenly deleted llama_batch_init/free

ngxson committed:
fix llama-run n_past

ngxson committed:
fix gemma3-cli

ngxson committed:
fix missing n_past in various places

this is actually a revert of https://github.com/ggml-org/llama.cpp/commit/cda0e4b648dde8fac162b3430b14a99597d3d74f

ngxson committed:
fix llama_batch_ext_init_from_embd

ngxson committed:
qwen2vl: use llama_batch_ext_set_pos

ngxson committed:
fix compile

ngxson commented:
@ggerganov This PR is ready for review now, everything is now working except for:
- `speculative.cpp` which I use `#if 0` is disable the code. I don't have enough knowledge for fix it, so can you please give a try?
- `android` and `swiftui`, I also don't have enough knowledge to code on swift and Java JNI, so better to leave them as-is for now and maybe wait for the community to fix it later on.

ggerganov commented:
Thanks. I'll take a look at the remaining examples.

ngxson committed:
llama_batch_ext_ptr::from_text/embd

ngxson committed:
rename to init_from_text

ngxson committed:
fix compile

ggerganov commented:
The `tts` example currently fails:

```bash
./bin/llama-tts --tts-oute-default -p "Hello world" -lv 1
```

```
0.05.452.175 E get_logits_ith: invalid logits id 0, reason: batch.logits[0] != true
Abort trap: 6
```

ggerganov commented on examples/tts/tts.cpp:
```diff
@@ -917,28 +917,28 @@ lovely<|t_0.56|><|code_start|><|634|><|596|><|1766|><|1556|><|1306|><|1285|><|14
 
                 // push this new token for next evaluation
-                common_batch_add(batch, new_token_id, n_past, { i }, true);
+                llama_batch_ext_add_text(batch, new_token_id, n_past, &i, 1, false);
```
```suggestion
                llama_batch_ext_add_text(batch, new_token_id, n_past, &i, 1, true);
```

ngxson committed:
Update examples/tts/tts.cpp

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

ngxson commented on examples/batched/batched.cpp:
```diff
@@ -193,23 +193,23 @@ int main(int argc, char ** argv) {
 
             // push this new token for next evaluation
-            common_batch_add(batch, new_token_id, n_cur, { i }, true);
+            llama_batch_ext_add_text(batch, new_token_id, n_cur, &i, 1, false);
```
```suggestion
            llama_batch_ext_add_text(batch, new_token_id, n_cur, &i, 1, true);
```

ngxson commented on examples/batched-bench/batched-bench.cpp:
```diff
@@ -150,10 +144,10 @@ int main(int argc, char ** argv) {
 
                     for (int j = 0; j < pl; ++j) {
-                        common_batch_add(batch, 0, pp + i, { j }, true);
+                        llama_batch_ext_add_text(batch, 0, pp + i, &j, 1, false);
```
```suggestion
                        llama_batch_ext_add_text(batch, 0, pp + i, &j, 1, true);
```

ngxson commented on examples/main/main.cpp:
```diff
@@ -668,7 +669,9 @@ int main(int argc, char ** argv) {
 
-                if (llama_decode(ctx, llama_batch_get_one(&embd[i], n_eval))) {
+                auto batch = llama_batch_ext_ptr::init_from_text(&embd[i], n_eval, n_past, 0, true);
+                llama_batch_ext_set_output_last(batch.get());
```
(already set via the last argument of `init_from_text`)

```suggestion
```

ngxson committed:
Apply suggestions from code review

ggerganov committed:
speculative : adapt to new llama API

ggml-ci

ngxson committed:
Merge branch 'master' into xsn/private_batch_api

ngxson committed:
Merge pull request #15 from ggml-org/xsn/private_batch_api

speculative : adapt to new llama API

ggerganov committed:
android : adapt to new API

ggerganov commented:
The `llama-android` example should be updated now, but I can't fully test it because something with the read permissions on Android has changed and it does not allow me to load the model in the simulator:

![image](https://github.com/user-attachments/assets/c3e4576b-49ec-4056-8e95-44fd324a38ae)

ngxson commented:
It can be because the permission dialog not being shown, have you checked in settings -> apps -> permissions and activate the file access permission manually for the app?

(I'll also try if I can free up some spaces on my macbook)

ggerganov committed:
swift : adapt to new API

ggerganov commented:
I think the app didn't request the permission, so it is not shown:

![image](https://github.com/user-attachments/assets/e365e27d-1dd9-44fc-be4c-a933299665d2)

ngxson commented:
Ok seems like it's missing the storage permission in manifest.xml file (and also the code to show permission dialog, but it's optional)

ggerganov commented:
> Never mind, I downloaded android studio and testing it. Seems like external storage access permission is completely removed in a newer android version. I'm trying with a local app data dir.

Yes, that's what I figured as well.

In the meantime, I will try to reduce the usage of `common_batch` to a minimum, and ideally remove it all together. Maybe only needed in `server`.

ngxson committed:
android : fix permission

ggerganov committed:
retrieval : avoid common_batch

ggml-ci

ggerganov committed:
embedding : avoid common_batch

ggml-ci

ngxson commented:
While we're at this, I have a small question regarding the design of batch/ubatch: Currently, when entering a list of tokens, user must firstly split the list into smaller batches, then llama_decode/encode further split it into smaller ubatches.

So I'm wondering if it is possible to design the llama_batch_ext in such a way that it contains all the input tokens (no more n_batch limit). If this is something preferable, we can implement this in a follow-up PR.

ggerganov commented:
> So I'm wondering if it is possible to design the llama_batch_ext in such a way that it contains all the input tokens (no more n_batch limit).

Hm, not sure. Can you give a specific example where this would be beneficial?

I think that all apps require to have some logical batching of the inputs. For example, in `llama-perplexity`, we don't want to submit the entire Wiki text in a single logical batch, because it would block the main program for the entire processing. If we do so, it would still work correctly, but we lose some flexibility of doing things like printing progress, saving checkpoints, etc.

ngxson commented:
What I mean is that currently we have both for `n_ubatch` and `n_batch` in the core library, while from an user perspective, `n_batch` is actually optional and only used if the user need it.

In the example that you gave, yes it is beneficial to have a `n_batch` limit, but I think this limit should be up to the user to decide, and the core library should not care about that.

On the other hand, in most CLI chat apps like llama-run, llama-cli, llama-simple-chat, this batching division actually introduce quite some redundant boilerplate code as we don't want the user to interrupt while it's processing (risk of resulting in a broken half-process context I think?)

ggerganov committed:
perplexity : avoid common_batch

ggml-ci

ggerganov commented:
> What I mean is that currently we have both for n_ubatch and n_batch in the core library, while from an user perspective, n_batch is actually optional and only used if the user need it.

Yes, I think I see your point. There seems to be no reason for the `libllama` to know about `n_batch`. It seems like the only place currently that utilizes it is the output reserve logic. But I think this code has to be reworked either way.

ggerganov commented:
Will try to avoid `common_batch` from the `llama-server` now.

ggerganov committed:
server : avoid common_batch

ggml-ci

ggerganov commented:
Opened a suggestion in https://github.com/ngxson/llama.cpp/pull/16.

ggerganov committed:
server : remove old commented code [no ci]

ngxson committed:
Merge pull request #16 from ggml-org/xsn/private_batch_api_pooling_none

server : avoid common_batch

ngxson commented:
Thanks for helping me to complete this PR, feel free to merge when you're ready.

Only my side, I tested these binaries and confirmed that it works: main.cpp, run, minicpmv-cli, server, perplexity

ggerganov commented on examples/batched/batched.cpp:
```diff
@@ -155,16 +155,16 @@ int main(int argc, char ** argv) {
     // remember the batch index of the last token for each parallel sequence
     // we need this to determine which logits to sample from
-    std::vector<int32_t> i_batch(n_parallel, batch.n_tokens - 1);
+    std::vector<int32_t> i_batch(n_parallel, llama_batch_ext_get_n_tokens(batch) - 1);
```
I think the next step along this refactoring is to remove all usages of `i_batch` from the examples. The `i_batch` is the index we use to extract logits from the `i`th token in the batch, but this pattern is quite cumbersome and not very intuitive. In order to avoid this pattern, we have to introduce a new API call for sampling a token from a sequence:

```c
    LLAMA_API llama_token llama_sampler_sample_seq(struct llama_sampler * smpl, struct llama_context * ctx, llama_seq_id seq_id);
```

This should be enough to replace most or all usages of `i_batch`. We can do this in a next PR.

ggerganov commented on include/llama.h:
```diff
@@ -902,13 +909,98 @@ extern "C" {
 
-    // Frees a batch of tokens allocated with llama_batch_init()
-    LLAMA_API void llama_batch_free(struct llama_batch batch);
+    // Same with llama_batch_init, but initializes the batch with the provided text tokens
+    // First token will be at position pos0
+    // The sequence ID will be fixed to seq_id
+    // If output_last is true, the last token will have output set
+    // The batch has to be freed with llama_batch_ext_free()
+    LLAMA_API struct llama_batch_ext * llama_batch_ext_init_from_text(
+            llama_token * tokens,
+                int32_t   n_tokens,
+                int32_t   pos0,
+                int32_t   seq_id,
+                   bool   output_last);
```
Since this call is basically a combination of other calls in the API, we should probably remove it in order to keep the API smaller. Moreover, we can use the C++ version `llama_batch_ext_ptr::init_from_text()`. The C++ API in `llama-cpp.h` can absorb most of the convenience API calls since it supports argument overloading.

ngxson replied:
Hmm yeah that makes sense. Indeed, I aimed to have this C API to be exposed to downstream binding, but on second thought it's better to keep the API minimal

ngxson replied:
Removed in https://github.com/ggml-org/llama.cpp/pull/11875/commits/30f1db9936112bc04917df1a9520421ae5f3ffc1

ggerganov commented on include/llama.h:
```diff
@@ -902,13 +909,98 @@ extern "C" {
+    // Get the number of tokens in the batch
+    LLAMA_API int32_t llama_batch_ext_get_n_tokens(const struct llama_batch_ext * batch);
+
+    // Add text tokens to the batch
+    // Return values:
+    // -1 : not enough space in the batch
+    // -2 : embd is already set, cannot add text tokens
+    // otherwise, returns the output ID
+    LLAMA_API int32_t llama_batch_ext_add_text(
+        struct llama_batch_ext * batch,
+                   llama_token   token,
+                     llama_pos   pos,
+            const llama_seq_id * seq_ids,
+                        size_t   n_seq_ids,
+                          bool   output);
```
This should probably support adding multiple tokens at once (i.e. make it more general). And we should add a C++ convenience wrapper that allows adding a single token with a single sequence id.

ngxson replied:
I'm thinking about adding a member function `llama_batch_ext_ptr::add_seq` that operates explicitly with `llama_batch_ext_ptr` and does not accept plain C pointer. But this will requires moving a lot of places from plain pointer to unique_ptr. Do you think this is expected?

ngxson replied:
Ok so I experimented this idea on `perplexity.cpp` and seems like it's more intuitive to create the wrapper for adding one single token instead of adding the whole sequence: https://github.com/ggml-org/llama.cpp/pull/11875/commits/2134cabf50ab327cbb9c18150cc4fbe5d2bc8d36

The `batch_add_seq` that you proposed is a bit overlap with `llama_batch_ext_ptr::init_from_text`. The main problem here is that in many cases, we want to add a subset of tokens from the prompt, something like this:

```
const int offset = ...;
for (size_t i = 0; i < some_number; ++i) {
    batch.add_text(tokens[i + offset], i + offset, 0, false);
}
```

So in this case I think it's preferable to keep the `for` loop. Otherwise, we can also add an overload for iterator, but the code will look quite the same as the `for` loop.

ggerganov replied:
This change looks good to me.

> So in this case I think it's preferable to keep the for loop.

Ok, agree.

ngxson replied:
(copying this comment here because I put it in the wrong place)

Ok I've been able to apply the change to various example, I leave some examples as-is for now since they use multi seq, thus the CPP API does not make it looks shorter. Here is the commit: https://github.com/ggml-org/llama.cpp/pull/11875/commits/2cec1cff7412999c1002949fc53edc2204361575

And going a bit further, I also added `batch.clear()` and `batch.n_tokens()` to make the code nicer to look at: https://github.com/ggml-org/llama.cpp/pull/11875/commits/3802ff2a9b8df77d28797d53eb3d9dca24f960d1

ggerganov reviewed (commented):
I think we can iterate a little bit more to make the usage of the API more convenient.

For example, this pattern of setting the sequence id is a bit annoying:

```c
llama_seq_id seq_id = 0;
llama_batch_ext_add_text(batch, id, pos, &seq_id, 1, true);
```

Also, adding multiple tokens to a batch should be simpler, because it is used often:

```c
static void batch_add_seq(llama_batch_ext * batch, const std::vector<int32_t> & tokens, llama_seq_id seq_id) {
    size_t n_tokens = tokens.size();
    for (size_t i = 0; i < n_tokens; i++) {
        llama_batch_ext_add_text(batch, tokens[i], i, &seq_id, 1, true);
    }
}
```

I think we can utilize the C++ API in `llama-cpp.h` to improve on these aspects. See the comments below.

ngxson commented on include/llama.h:
```diff
@@ -902,13 +909,98 @@ extern "C" {
+    // Note: this is only to be used in conjunction with llama_batch_ext_init_from_embd()
+    // n_pos must match the n_tokens of the batch
+    // Returns -1 if n_pos does not match the n_tokens of the batch
+    LLAMA_API int32_t llama_batch_ext_set_pos(struct llama_batch_ext * batch, llama_pos * pos, size_t n_pos);
```
I thought about that, but the problem is that I can't come up with an intuitive signature for `llama_batch_ext_init_from_embd` in that case.

My first idea was:

```cpp
    LLAMA_API struct llama_batch_ext * llama_batch_ext_init_from_embd(
              float * embd,
             size_t   n_tokens,
             size_t   n_embd,
          llama_pos * pos,
             size_t * n_pos
            int32_t   seq_id)
```

But the problem was that for all the code using "normal" position, this introduce more boilerplate code.

Another idea is to have separated calls:
- `llama_batch_ext_init_from_embd` that takes `pos0`
- `llama_batch_ext_init_from_embd_custom_pos` that takes an array of `pos` and array length `n_pos`, but I can't thnk of a better naming

We can also have both `pos0`, `pos` and `n_pos` on the same function, but this will be quite confused for develop IMO.

So for now having a separated call `llama_batch_ext_set_pos` works best, we can also reuse it in a future `llama_batch_ext_init_from_tensor`. WDYT?

ngxson committed:
remove C API llama_batch_ext_init_from_text

ngxson committed:
Merge branch 'master' into xsn/private_batch_api

ngxson committed:
add cpp batch.add_text wrapper

ggerganov commented on src/llama-batch.cpp:
```diff
@@ -323,6 +337,167 @@ struct llama_batch llama_batch_get_one(
+    if ((size_t) batch->n_tokens != n_pos) {
+        return -1;
+    }
+    memcpy(batch->pos, pos, n_pos * sizeof(llama_pos));
```
I think this will copy only 1/4 of the positions for Qwen2VL.

ngxson replied:
Edit: on closer look, seems like this is intentional. `image_embed` is divided into 4 batches hence only 1/4 positions being used. For now, I have no idea why.

ggerganov replied:
I think because on `master` the batch stores a pointer to a buffer with positions, it works correctly thanks to the `n_pos_per_token == 4`:

https://github.com/ggml-org/llama.cpp/blob/30f1db9936112bc04917df1a9520421ae5f3ffc1/src/llama-graph.cpp#L54-L60

But now this will likely do out-of-bounds read because we allocate a smaller buffer.

It's probably a good idea to figure this out, because it can change the API significantly. For example, `llama_batch_ext_init()` might need to accept `n_pos_per_token` argument (or a `llama_model` reference) so that it can allocate the appropriate `pos` array.

ngxson replied:
Oh ok I didn't know that qwen2vl needs multiple pos per token.

Since qwen2vl is the only model supporting this, I think we should not add `n_pos_per_token` directly to `llama_batch_ext_init()`, this can make the API looks quite confusing.

Instead, we can simply allocate `batch.tokens` to `MAX_POS_PER_TOKEN * n_tokens` with `MAX_POS_PER_TOKEN` fixed to 4. In the future when the vision API is added, we could probably hide this position calculation into `libllama`. My idea looks like this:

```cpp
llama_mm_embd embd = llama_vision_get_output_embd(ctx);
llama_batch_ext * batch = llama_batch_ext_from_mm_embd(embd); // the mrope position is calculated here
```

ngxson replied:
I proposed a fix in https://github.com/ggml-org/llama.cpp/pull/11875/commits/a9efdbbce56e95579b255d0052b289a76c1e7d9f

But this qwen2vl example seems to be broken even on `master`. I asked it to describe the windows xp bliss wallpaper and it gives wrong answers. Probably better to skip this and redo the whole thing on my vision refactoring.

ngxson committed:
move various places to batch.add_text

ngxson committed:
add batch.clear() and batch.n_tokens()

ngxson committed:
Merge branch 'master' into xsn/private_batch_api

ngxson committed:
qwen2vl: fix mrope position

ngxson commented:
@ggerganov Could you take a look on this PR this week? Thanks!

ggerganov reviewed (commented):
Sorry for the delay. I still have some concerns:

- I am thinking that it would be a good idea to make the `llama_batch_ext` API to treat tokens and embeddings in a very similar manner. Ultimately, I think we should be able to create batches that contain both tokens and embeddings. For example the call:

```c
    LLAMA_API struct llama_batch_ext * llama_batch_ext_init(
            int32_t n_tokens,
            int32_t n_seq_max);
```

might be better to define as:

```c
    // either one of these - not sure which one yet
    LLAMA_API struct llama_batch_ext * llama_batch_ext_init(
            struct llama_model * model,
            int32_t n_seq_max);

    // this one will figure out `n_seq_max` from the context
    // maybe actually this one is the best
    LLAMA_API struct llama_batch_ext * llama_batch_ext_init(
            struct llama_context * ctx);
```

Passing `n_tokens` to the batch init call was necessary in the past in order to pre-allocate an array with enough size. But it is technically redundant information, because we can add new tokens and embeddings and dynamically resize the batch in `libllama` as needed. So I think there is no longer need to provide this information.

- I fill like the embeddings API should mirror the tokens API. So instead of:

```c
    LLAMA_API struct llama_batch_ext * llama_batch_ext_init_from_embd(
                  const float * embd,
                       size_t   n_tokens,
                       size_t   n_embd,
                    llama_pos   pos0,
                 llama_seq_id   seq_id);

    LLAMA_API int32_t llama_batch_ext_set_pos(struct llama_batch_ext * batch, llama_pos * pos, size_t n_pos);
```

we should probably have something like:

```c
    LLAMA_API int32_t llama_batch_ext_add_embd(
        struct llama_batch_ext * batch,
                   const float * embd, // size would be inferred from the context
                     llama_pos * pos,  // multiple pos per embd
                        size_t   n_pos,
            const llama_seq_id * seq_ids,
                        size_t   n_seq_ids,
                          bool   output);
```

I think these change would help us think about tokens and embeddings as something very similar.

ngxson commented:
Nice, thanks for the comment. Yes I agree with your points. For `llama_batch_ext_init` it should take the `llama_context * ctx` because I think one batch should only be valid for a given context (in case we create multiple contexts from single model). I'll work on this now.

ngxson committed:
Merge branch 'master' into xsn/private_batch_api

ngxson committed:
llama_batch_ext_init with ctx

ngxson commented:
Ok so I implemented the new `llama_batch_ext_init` and `llama_batch_ext_add_embd` in my last commit: https://github.com/ggml-org/llama.cpp/pull/11875/commits/d18a79ed07c61e21474db1097ccf929b98f044e7

For now, qwen2vl-cli is broken because the `llama_batch_ext_add_text` API only accepts one single position. We can implement the multi pos per token inside this API itself, but a little bit tricky. Let's take an example: we have `n_pos_per_token == 4`, then we add tokens at position 1, 2, 3, 4, 5, 6, we expect the `pos` array to be:

`111122223333444455556666`

But in reality, cgraph only accepts:

`123456123456123456123456`

One is the transposed version of another, so will be simple to do: before `llama_de/encode` we can reorder the `pos` so that it has the correct layout.

But before implementing this, I just want to check with you if my direction still looks ok.

ggerganov commented on src/llama-batch.cpp:
```diff
@@ -370,34 +370,28 @@ static struct llama_batch_ext * llama_batch_ext_init_impl(int32_t n_tokens_alloc
-struct llama_batch_ext * llama_batch_ext_init(int32_t n_tokens_alloc, int32_t n_seq_max) {
-    return llama_batch_ext_init_impl(n_tokens_alloc, 0, n_seq_max);
+struct llama_batch_ext * llama_batch_ext_init(struct llama_context * ctx) {
+    return llama_batch_ext_init_impl(llama_n_batch(ctx), 0, llama_n_seq_max(ctx));
```
Yes, for now this is a good solution. Later, we will be resizing dynamically and not need the `llama_n_batch()`.

ggerganov commented:
> But before implementing this, I just want to check with you if my direction still looks ok.

I think this looks good.

ngxson committed:
fix qwzn2vl mrope position input

ngxson committed:
fix build

ngxson commented:
I implemented the fix, tested `qwen2vl-cli` and it seems to work with text. Can't test with image for now since it's also broken on `master`

The command is: `llama-qwen2vl-cli -m ../models/Qwen2-VL-7B-Instruct-Q4_K_M.gguf --mmproj ../models/mmproj-Qwen2-VL-7B-Instruct-f16.gguf --image ../models/bliss.png -p "what do you see?"`

---

A bit confused about this, I found this nice illustration on qwen2vl model page on HF, which shows that qwen only use 3 pos per token. It's also confirmed by the [config.json](https://huggingface.co/Qwen/Qwen2-VL-7B-Instruct/blob/main/config.json#L45) file. However, I'm not sure why in llama.cpp we use up to 4 pos.

<img width="1385" alt="image" src="https://github.com/user-attachments/assets/f9fe7c5d-3376-4798-b622-46a6a9e51561" />

ggerganov commented:
> bit confused about this, I found this nice illustration on qwen2vl model page on HF, which shows that qwen only use 3 pos per token. It's also confirmed by the [config.json](https://huggingface.co/Qwen/Qwen2-VL-7B-Instruct/blob/main/config.json#L45) file. However, I'm not sure why in llama.cpp we use up to 4 pos.

cc @HimariO

ngxson committed:
fix server

ngxson committed:
server: fix batch_spec

ggerganov commented on examples/server/server.cpp:
```diff
@@ -1963,7 +1963,7 @@ struct server_context {
             const int32_t n_batch = llama_n_batch(ctx);
 
             // only a single seq_id per token is needed
```
This comment is obsolete.

ngxson committed:
fix embeddings and retrieval

ngxson committed:
correct output_id for llama-cpp header

HimariO commented:
> A bit confused about this, I found this nice illustration on qwen2vl model page on HF, which shows that qwen only use 3 pos per token. It's also confirmed by the [config.json](https://huggingface.co/Qwen/Qwen2-VL-7B-Instruct/blob/main/config.json#L45) file. However, I'm not sure why in llama.cpp we use up to 4 pos.
The fourth position ID is mainly for future-proofing, in case a newer model that takes 3D/depth input(like SpatialLM) is added. Currently, both Qwen2 & 2.5VL only use 3 position ID per token.

ggerganov commented:
@ngxson We discussed with @slaren these changes and he raised a good point that the batch API does not need to explicitly pass token positions. These can be inferred from the KV cache.

Since not having to pass explicit token positions would simplify the batch API, it's a good idea to take it into account when redesigning it. So I will try to do some KV cache refactoring to support this. When I'm ready, I will come back to this PR and update it respectively.

ggerganov commented:
I think the Qwen2VL image positions could be inferred from the token position:

https://github.com/ggml-org/llama.cpp/blob/3fd072a54001a908c54e81fd2e82b682ecfdd475/examples/llava/qwen2vl-cli.cpp#L39-L50

So ideally, the user would not have to pass those as well.

ngxson commented:
We still need to know the image size `pw` and `ph` in order to infer the 2D position. Ofc my goal for the future is to hide all of this inside the library. This is the idea in [this comment](https://github.com/ggml-org/llama.cpp/pull/11875#discussion_r2011880062) where I suggest having a `struct llama_mm_embd` that contains the image embeddings tensor and the size of image `pw`, `ph`

ggerganov commented:
Yes, the multi-dim positions are complicating things. Not sure what is the best solution.

> The problem is that many vision models nowadays also need to know the size of image when passing the embeddings from encoder to decoder.

Maybe with Gemma demonstrating that this is not necessary, new models won't need these complications and we don't have to add support at all. Are there other models other than Qwen2VL that need 2D positions?

> When passing these embeddings to decoder, we need to "wrap" each slice between some tokens, then also "wrap" each row of 3 slices between another token:

This seems something that the user code can implement logic (similar how to bos/eos tokens are added). Which models use this pattern?

Anyway, the KV cache refactoring can be done before making a decision about how to handle the images, so we can re-discuss this after that.

ngxson commented:
Just to clarify before writing my response, there are 2 reasons why the image size is needed:
- For 2D positional embedding like Qwen, where each token has 3 positions (temporal, X, Y)
- For slices layout --> currently use by a lot of models like MiniCPM-V, SmolVLM, llava 1.6. This technique use normal position and normal causal mask, but they add special tokens to identify rows/cols of slices

> Maybe with Gemma demonstrating that this is not necessary, new models won't need these complications and we don't have to add support at all. Are there other models other than Qwen2VL that need 2D positions?

Hmm yeah that could be right. Because M-RoPE is invented by Qwen, I don't see anyone using it for now. Not sure if other models will adopt it in the future.

But please not that gemma 3 does not use slices. This makes working with gemma 3 vision easy, but the current problem with gemma 3 is that the image size is fixed. For bigger images, they need to rely on a technique called "pan and zoom", which essentially a prompting technique that allow the model to "ask" the runtime to zoom the image, then rerun the generation. This is obviously very inefficient.

Models like SmolVLM, MiniCPM-V, Qwen (and maybe many other) are already using slicing technique that I said earlier, so we definitely need to support this in the API.

> This seems something that the user code can implement logic (similar how to bos/eos tokens are added). Which models use this pattern?

In fact, better to think is that is the "chat template" for image. While it can be implemented in user code, I think it's better to make it transparent from user POV, as this part is model-specific and even harder than normal chat template for text, this is not something user can easily debug.

In gemma3-cli, you can see that the `<start_of_image>` token is added from user code. But my goal for the vision API is to hide this behind an API.

ngxson commented:
@ggerganov I've been working on audio [input](https://github.com/ggml-org/llama.cpp/pull/12745) and [output](https://github.com/ggml-org/llama.cpp/pull/12648) recently and I think the API proposed by this PR pretty much correspond to what I need (ofc except for the position, which will be nicer to be hidden from user code).

Having this PR merged could save me some efforts, and more importantly unblock my researches on multimodal API, so I'm wondering if anything I could do on my side to accelerate this a bit more? Thank you!

ggerganov commented:
My plan for the next steps was to refactor the `llama_kv_cache_unified` into 2 separate implementations - unified and recurrent (started some initial work in #12695). This will simplify the logic in `llama_kv_cache` and allow to design a better token position / sequence tracking in `libllama` which can then be used instead of manually passing positions with the batches. When this is ready, we can update this PR to not pass the positions explicitly.

> Also now that the position is hidden from user code, I think we should also somehow modify the llama_kv_* API to make sure that user don't accidentally leave a "hole" in the context. For example, if KV cache has 10 tokens and they delete tokens from position [2, 5), now there will be no API to fill in this "hole" unless we also remove all tokens [5, inf)

The `llama_kv_self_seq_` API would still require the user to keep track of the token positions / sequence lengths. But not sure if this can be improved.

ngxson commented:
@ggerganov I'm re-designing the API to take into account the MTP embeddings and `llama_process()`, coming up with this. WDYT?

I added `llama_process_type` to be either ENCODE/DECODE or something else in the future, but probably I misunderstood your intention.

```cpp

    //
    // Extended batch API
    //

    struct llama_batch_ext;

    LLAMA_API struct llama_batch_ext * llama_batch_ext_init (struct llama_context * ctx);
    LLAMA_API void                     llama_batch_ext_free (struct llama_batch_ext * batch);
    LLAMA_API void                     llama_batch_ext_clear(struct llama_batch_ext * batch);

    struct llama_batch_token {
        llama_token     id;         // if id != LLAMA_TOKEN_NULL, embd must be nullptr
        float         * embd;       // if embd != nullptr, token id must be LLAMA_TOKEN_NULL
        float         * embd_nextn; // used by nextn layers
        llama_pos     * pos;        // if nullptr, the position will be automatically assigned
                                    // for M-RoPE models, embedding tokens must have multiple positions per token; text token only requires one single position per token
        llama_seq_id    seq_id;
        bool            output;
    };

    // Add an input token to the batch
    // Returns the batch index (>= 0)
    // On error:
    //    -1: batch is full
    //    -2: token is invalid (id == LLAMA_TOKEN_NULL or invalid embd)
    //    -3: invalid sequence id
    LLAMA_API int32_t llama_batch_ext_add_token(
            struct llama_batch_ext * batch,
                 llama_batch_token   token);

    // Set output = true for the last added token in the batch
    LLAMA_API int32_t llama_batch_ext_set_output_last(
            struct llama_batch_ext * batch);

    // Return values are the same as llama_decode()
    LLAMA_API int32_t llama_process(
                struct llama_context * ctx,
             enum llama_process_type   type,
              struct llama_batch_ext * batch);

```

Basic usage:

```cpp
batch = llama_batch_ext_init(ctx);
for (auto token_id : input_tokens) {
  llama_batch_token t{
    token_id,
    nullptr, // embd
    nullptr, // embd_nextn
    nullptr, // pos
    0, // seq_id
    false // output
  };
}
llama_batch_ext_set_output_last(batch);
llama_process(ctx, LLAMA_PROCESS_TYPE_DECODE, batch);
```

ngxson commented:
I moved this PR to https://github.com/ggml-org/llama.cpp/pull/24669
